lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
@aprescott Thanks for the tip, but ConnectionPool still doesn't help us there with version 2.1.3 as we still get errors, though slightly different ones:

`ActiveRecord::StatementInvalid: Mysql2::Error: Lost connection to MySQL server during query: SELECT ...`